File: binary_search.rb
def search(arr, key)
   lo = 0
   hi = arr.length 
   while (lo <= hi) 
      mid = lo + (hi - lo) / 2
      if (key < arr[mid])
         hi = mid - 1
      elsif (key > arr[mid])
         lo = mid + 1
      else
         return mid 
      end
   end 
   return -1
end 

arr = [2, 3, 5, 8, 11, 30, 55, 88, 233, 534, 553]
puts search(arr, 11)
puts search(arr, 233)
puts search(arr, 12)

File: BinarySearch.swift
// This method will return index of array if search item is found
// Otherwise it will return -(insertion position + 1)
func binarySearch<T:Comparable>(inputArr:Array<T>, searchItem: T)->Int{
    return binarySearch0(inputArr, fromIndex: 0, toIndex: inputArr.count, searchItem: searchItem)
}

private func binarySearch0<T:Comparable>(inputArr:Array<T>, fromIndex: Int, toIndex: Int, searchItem: T)-> Int {
    var low = fromIndex
    var high = toIndex - 1
    
    while (low <= high) {
        let mid = low + (high - low) / 2
        let midVal = inputArr[mid]
        
        if (midVal < searchItem) {
            low = mid + 1
        } else if midVal > searchItem {
            high = mid - 1
        } else {
            return mid // key found
        }
    }
    return -(low + 1)  // key not found.
}
